### PR TITLE
Bugfix/check for approve reject

### PR DESCRIPTION
**Pull request by serjKim**

### CODE REVIEW

File: server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs [hunk 1]
```diff
@@ -162,7 +162,16 @@ public CalendarEventsController(ITimeoutSettings timeoutSettings, IEmployeesRegi
                 return this.NotFound();
             }
 
-            if (!(await this.authorizationService.AuthorizeAsync(this.User, employee, new ApproveCalendarEvents())).Succeeded)
+            var approvedStatus = new CalendarEventStatuses().ApprovedForType(model.Type);
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs]: can this logic be moved on Requirement level?

Comment (santee) [reply server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs]: also, this probably should be 'EditCalendarEvents' and 'EditPendingCalendarEvents'. That way we also check date changes

File: server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs [hunk 2]
```diff
@@ -162,7 +162,12 @@ public CalendarEventsController(ITimeoutSettings timeoutSettings, IEmployeesRegi
                 return this.NotFound();
             }
 
-            if (!(await this.authorizationService.AuthorizeAsync(this.User, employee, new ApproveCalendarEvents())).Succeeded)
+            if (!(await this.authorizationService.AuthorizeAsync(this.User, employee, new ApproveCalendarEvents(model))).Succeeded)
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs]: let's have two permissions instead - EditCalendarEvents, EditDraftCalendarEvents.

**Comment by serjKim** [comment]
Updated

File: server/Arcadia.Assistant.Security/PermissionsActor.cs [hunk 3]
```diff
@@ -182,15 +182,20 @@ private void ReplyAndStop()
 
         private static readonly EmployeePermissionsEntry SelfPermissions =
             EmployeePermissionsEntry.CreateCalendarEvents |
+            EmployeePermissionsEntry.CompleteCalendarEvents |
```
Comment (santee) [inline server/Arcadia.Assistant.Security/PermissionsActor.cs]: Care to name it as it is? ProlongSickLeave | CompleteSickLeave?

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCommonCalendarEventsStatusesPermissionHandler.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            {
+                if (employeePermissions.HasFlag(EmployeePermissionsEntry.RejectCalendarEvents))
+                {
+                    context.Succeed(requirement);
+                }
+
+                return;
+            }
+
+            context.Succeed(requirement);
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCommonCalendarEventsStatusesPermissionHandler.cs]: that's weird. What if I do not have permissions to edit status? :)

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            var employeePermissions = allPermissions.GetPermissions(resource);
+
+            var existingEvent = requirement.ExistingEvent;
+            var updatedEvent = requirement.UpdatedEvent;
+
+            var calendarEventStatuses = new CalendarEventStatuses();
+            var approved = calendarEventStatuses.ApprovedForType(existingEvent.Type);
+            var rejected = calendarEventStatuses.RejectedForType(existingEvent.Type);
+
+            if (updatedEvent.Status != existingEvent.Status && updatedEvent.Status == approved)
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: put it in a variable statusChanged?

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+            if (updatedEvent.Type == CalendarEventTypes.Sickleave)
+            {
+                if (updatedEvent.Status == existingEvent.Status
+                    && updatedEvent.Status == SickLeaveStatuses.Approved
+                    && updatedEvent.Dates != existingEvent.Dates)
+                {
+                    if (employeePermissions.HasFlag(EmployeePermissionsEntry.ProlongSickLeave))
+                    {
+                        context.Succeed(requirement);
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: nah , you'd need some failures. If it's false but the next is true you get success. I'd just got rid of return calls and went if Succeed/Failure calls on permissions. Also, put each condition in a separate method maybe?

File: server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs [hunk 7]
```diff
@@ -162,17 +162,21 @@ public CalendarEventsController(ITimeoutSettings timeoutSettings, IEmployeesRegi
...
-            }
-
             var existingEvent = await this.GetCalendarEventOrDefaultAsync(employee.Calendar.CalendarActor, eventId, token);
             if (existingEvent == null)
             {
                 return this.NotFound();
             }
 
+            var hasPermissions = existingEvent.IsPending
+                ? (await this.authorizationService.AuthorizeAsync(this.User, employee, new EditPendingCalendarEvents())).Succeeded
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Controllers/CalendarEventsController.cs]: var requirement = existingEvent.isPending ? new EditPending... : new Edit; await AuthorizeAsync

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                hasPermissions &= this.CheckSickLeave(existingEvent, updatedEvent, employeePermissions);
+            }
+
+            if (hasPermissions)
+            {
+                context.Succeed(requirement);
+            }
+            else
+            {
+                context.Fail();
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: maybe get rid of else block? if the requirement is not satisfied by this handler it still can be satisfied with any other

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            }
+        }
+
+        private bool CheckIfApproval(bool statusChanged, CalendarEvent existingEvent, CalendarEventsModel updatedEvent, EmployeePermissionsEntry employeePermissions)
+        {
+            var calendarEventStatuses = new CalendarEventStatuses();
+            var approved = calendarEventStatuses.ApprovedForType(existingEvent.Type);
+
+            if (statusChanged && updatedEvent.Status == approved)
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: what do you think if we have statusChanged calculation in a separate method? I do not like that it's passed as a parameter...

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+            {
+                return employeePermissions.HasFlag(EmployeePermissionsEntry.RejectCalendarEvents);
+            }
+
+            return true;
+        }
+
+        private bool CheckSickLeave(CalendarEvent existingEvent, CalendarEventsModel updatedEvent, EmployeePermissionsEntry employeePermissions)
+        {
+            if (updatedEvent.Status == existingEvent.Status
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: 'statusChanged' :)

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+                && updatedEvent.Status == SickLeaveStatuses.Approved
+                && updatedEvent.Dates != existingEvent.Dates)
```
Comment (santee) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: are you sure that 'Dates' comparison actually calls Equals overload?

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs (hunk 9 above)
Comment (serjKim) [reply server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: i don't mind

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs (hunk 11 above)
Comment (serjKim) [reply server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: yes. Checked it with debug

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs (hunk 8 above)
Comment (serjKim) [reply server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: right

File: server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs [hunk 12]
```diff
@@ -40,16 +39,13 @@ public EditCalendarEventsPermissionHandler(IPermissionsLoader permissionsLoader)
...
                 context.Succeed(requirement);
             }
-            else
-            {
-                context.Fail();
-            }
         }
 
-        private bool CheckIfApproval(bool statusChanged, CalendarEvent existingEvent, CalendarEventsModel updatedEvent, EmployeePermissionsEntry employeePermissions)
+        private bool CheckIfApproval(CalendarEvent existingEvent, CalendarEventsModel updatedEvent, EmployeePermissionsEntry employeePermissions)
```
Comment (serjKim) [inline server/Arcadia.Assistant.Web/Authorization/Handlers/EditCalendarEventsPermissionHandler.cs]: maybe mark that method and others as static?

**Pull request merged**